django/db/models/sql/compiler.py

```diff
@@ -1452,6 +1454,24 @@ def single_alias(self):
         self.query.get_initial_alias()
         return sum(self.query.alias_refcount[t] > 0 for t in self.query.alias_map) == 1
 
+    def _expr_contains_self_ref_subquery(self, expr):
+        if hasattr(expr, 'rhs') and isinstance(expr.rhs, Subquery) and expr.rhs.query.model == self.query.model:
+            return True
+        if hasattr(expr, 'lhs') and isinstance(expr.lhs, Subquery) and expr.lhs.query.model == self.query.model:
+            return True
+        if hasattr(expr, 'children'):
+            return any(
+                self._expr_contains_self_ref_subquery(node)
+                for node in expr.children
+            )
+
+    @cached_property
+    def contains_self_reference_subquery(self):
+        return any(
+            self._expr_contains_self_ref_subquery(expr)
+            for expr in chain(self.query.annotations.values(), self.query.where.children)
+        )
+
```

Comment:
I didn't test it but I think this can be simplified to

```suggestion
    @classmethod
    def _expr_refs_alias(cls, expr, alias):
        if isinstance(expr, Query):
            return expr.get_initial_alias() == alias
        get_source_expressions = getattr(expr, 'get_source_expressions', None)
        if not get_source_expressions:
            return False
        return any(
            cls._expr_refs_alias(source_expr, alias) for source_expr in expr.get_source_expressions()
         )

    @cached_property
    def contains_self_reference_subquery(self):
        initial_alias = self.query.get_initial_alias()
        return any(
            self._expr_refs_alias(expr, initial_alias)
            for expr in chain(self.query.annotations.values(), self.query.where.children)
        )

```

Reply:
Thanks :+1:, I've changed  `_expr_refs_alias()` to the `_expr_refs_base_model()`, because they use a different aliases (e.g. `U0`, `admin_changelist_concert`) but the same model.